### PR TITLE
GraphQL server using Hot Chocolate and NHibernate

### PR DESCRIPTION
**Pull request by satishviswanathan**
**New Feature**

- NHibernate ORM integration 
- With this we can write database queries with LINQ.
- Add the ability to register NHibernate factory which can scan and registers all the NHibernate mapping files
- The library uses Fluent NHibernate so that we do have to deal with hbm.xml mapping files rather provide mapping in C# class files.
- This library has tried to follow the standards of EntityFrameWork library.

### CODE REVIEW

**Comment by michaelstaib** [comment]
Hi there, thanks for contributing! @PascalSenn will review this PR. We have tagged it for the October release. At the moment we are finishing version 12 of so it might take one or two weeks for the team to review this.

**Comment by satishviswanathan** [comment]
@PascalSenn Please do let me know if there is any to be done from my end.

**Comment by satishviswanathan** [comment]
@michaelstaib I've made changes for the public API's Shipped and pushed. Please let me know if it's good now

File: src/HotChocolate/Data/src/NHibernate/Extensions/NHibernateEnumerableExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+using System.Linq;
+
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/src/NHibernate/Extensions/NHibernateEnumerableExtensions.cs]:
```suggestion

```

File: src/HotChocolate/Data/src/NHibernate/Extensions/NHibernateFactory.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+//  any means, electronic, mechanical or otherwise, is prohibited without the
+//  prior written consent of the copyright owner.
+// </copyright>
+// 
+// <application>All</application>
+// <module>HotChoclate.Data.NHibernate</module>
+// <author>Viswanathan, Satish</author>
+// <createddate>2021-09-02</createddate>
+// <lastchangedby>Viswanathan, Satish</lastchangedby>
+// <lastchangeddate>2021-09-02</lastchangeddate>
```
Comment (michaelstaib) [inline src/HotChocolate/Data/src/NHibernate/Extensions/NHibernateFactory.cs:14]: We can only accept code that does not have copyright. I guess this is just a standard template where you did forget to clean.

```suggestion
```

File: src/HotChocolate/Data/src/NHibernate/Extensions/NHibernateFactory.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+           if (createSchema)
+           {
+               ISession session = sessionFactory.OpenSession();
+               new SchemaExport(configuration).Execute(true, true, false, session.Connection, Console.Out);
+
+               serviceCollection.AddSingleton(_ => session);
+           }
+           else
+           {
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/src/NHibernate/Extensions/NHibernateFactory.cs:61]:
```suggestion
```

File: src/HotChocolate/Data/src/NHibernate/Extensions/NHibernateFactory.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            {
+                cfg.FluentMappings.AddFromAssembly(assemblyContainingMapping);
+            })
+                .ExposeConfiguration(cfg => new SchemaExport(cfg).Create(true, true))
+                .BuildConfiguration();
+
+           ISessionFactory sessionFactory = configuration.BuildSessionFactory();
+           serviceCollection.AddSingleton(_ => sessionFactory);
+
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/src/NHibernate/Extensions/NHibernateFactory.cs:51]:
```suggestion

```

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/Author.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+
+
+namespace HotChocolate.Data
+{
+    public class Author
+    {
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/Author.cs:9]:
```suggestion
```

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/Author.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/Author.cs:4]:
```suggestion

```

File: src/HotChocolate/Data/src/NHibernate/publish.cmd [hunk 7]
```diff
@@ -0,0 +1,4 @@
+cls
```
Comment (michaelstaib) [inline src/HotChocolate/Data/src/NHibernate/publish.cmd:1]: We need to remember to remove this file before it gets merged. Our script files are located in the root.

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/AuthorFixture.cs [hunk 8]
```diff
@@ -0,0 +1,63 @@
+namespace HotChocolate.Data
+{
+    using System;
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/AuthorFixture.cs:3]: As mentioned above about usings.

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/AuthorFixture.cs [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        public IQueryable<ZeroAuthor> ZeroAuthor => _session.Query<ZeroAuthor>();
+        public IQueryable<SingleOrDefaultAuthor> SingleOrDefaultAuthor => _session.Query<SingleOrDefaultAuthor>();
+
+        public void Dispose()
+        {
+            File.Delete(_fileName);
+        }
+    }
+
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/AuthorFixture.cs:57]:
```suggestion

```

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/AuthorFixture.cs [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        private readonly string _fileName;
+
+        private readonly ISession _session;
+
+        public AuthorFixture() : base(typeof(Author).Assembly)
+        {
+            _fileName = Guid.NewGuid().ToString("N") + ".db";
+
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/AuthorFixture.cs:30]:
```suggestion

```

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/AuthorMap.cs [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+//  any means, electronic, mechanical or otherwise, is prohibited without the
+//  prior written consent of the copyright owner.
+// </copyright>
+// 
+// <application>All</application>
+// <module>HotChoclate.Data.NHibernate.Tests</module>
+// <author>Viswanathan, Satish</author>
+// <createddate>2021-09-01</createddate>
+// <lastchangedby>Viswanathan, Satish</lastchangedby>
+// <lastchangeddate>2021-09-01</lastchangeddate>
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/AuthorMap.cs:14]: As mentioned above about copyright.

```suggestion
```

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/BookMap.cs [hunk 12]
```diff
@@ -0,0 +1,23 @@
+
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/BookMap.cs:2]:
```suggestion
```

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/BookMap.cs [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+namespace HotChocolate.Data
+{
+    using FluentNHibernate.Mapping;
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/BookMap.cs:5]: As mentioned above on usings.

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/BookMap.cs [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+    public class BookMap : ClassMap<Book>
+    {
+
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/BookMap.cs:10]:
```suggestion
```

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/BookMap.cs [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+        {
+            Id(x => x.Id).Column("Id");
+            HasOne(x => x.Author).Cascade.All().PropertyRef(x=> x.Id);
+            Map(x => x.AuthorId);
+            Map(x => x.Title);
+            Table("Book");
+        }
+
+      
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/BookMap.cs:21]:
```suggestion
```

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/HotChoclate.Data.NHibernate.Tests.csproj [hunk 16]
```diff
@@ -0,0 +1,31 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/HotChoclate.Data.NHibernate.Tests.csproj:2]: @PascalSenn @michaelstaib we need to clean the project files and align them with the project.

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/InMemoryDatabaseTest.cs [hunk 17]
```diff
@@ -0,0 +1,48 @@
+
+namespace HotChocolate.Data
+{
+    using System;
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/InMemoryDatabaseTest.cs:4]: As mentioned above about usings.

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/InMemoryDatabaseTest.cs [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+        protected ISession Session;
+
+        public InMemoryDatabaseTest(Assembly assemblyContainingMapping)
+        {
+            if (_configuration == null)
+            {
+                FluentConfiguration fluentConfig = Fluently.Configure()
+                    .Database(SQLiteConfiguration.Standard.InMemory);
+
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/InMemoryDatabaseTest.cs:25]:
```suggestion

```

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/IntegrationTests.cs [hunk 19]
```diff
@@ -0,0 +1,437 @@
...
+using System.Threading.Tasks;
+using HotChocolate.Execution;
+using HotChocolate.Types;
+using Microsoft.Extensions.DependencyInjection;
+using Snapshooter.Xunit;
+using Xunit;
+
+namespace HotChocolate.Data
+{
+    using System.Linq;
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/IntegrationTests.cs:11]: As mentioned above about usings.

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/NHibernateExecutableTests.cs [hunk 20]
```diff
@@ -0,0 +1,74 @@
+namespace HotChocolate.Data
+{
+    using System.Linq;
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/NHibernateExecutableTests.cs:3]: As mentioned above about usings.

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/NHibernateExecutableTests.cs [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+    public class NHibernateExecutableTests
+    {
+        private readonly IQueryable<Author> _authors;
+
+        public NHibernateExecutableTests(AuthorFixture authorFixture)
+        {
+            _authors = authorFixture.Authors;
+        }
+
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/NHibernateExecutableTests.cs:18]:
```suggestion

```

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/NHibernateExecutableTests.cs [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+        [Fact]
+        public void Extensions_Should_ReturnEntityFrameworkExecutable_When_Queryable()
+        {
+            // arrange
+            // act
+            IExecutable<Author> executable = _authors
+                .AsQueryable()
+                .AsNhibernateExecutable();
+
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/NHibernateExecutableTests.cs:28]:
```suggestion

```

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/Query.cs [hunk 23]
```diff
@@ -0,0 +1,127 @@
+using System.Linq;
+using System.Threading.Tasks;
+using HotChocolate.Types;
+
+
+namespace HotChocolate.Data
+{
+    using NHibernate;
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/Query.cs:8]: As mentioned above about usings.

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/UseNHibernateSessionTests.cs [hunk 24]
```diff
@@ -0,0 +1,462 @@
+namespace HotChocolate.Data
+{
+    using System;
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/UseNHibernateSessionTests.cs:3]: As mentioned above about usings.

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/UseNHibernateSessionTests.cs [hunk 25]
```diff
@@ -0,0 +1,462 @@
...
+                    .AddProjections()
+                    .AddQueryType<Query>()
+                    .Services
+                    .BuildServiceProvider();
+
+            IRequestExecutor executor =
+                await services.GetRequiredService<IRequestExecutorResolver>()
+                    .GetRequestExecutorAsync();
+
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/UseNHibernateSessionTests.cs:40]:
```suggestion

```

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/UseNHibernateSessionTests.cs [hunk 26]
```diff
@@ -0,0 +1,462 @@
...
+    using Snapshooter.Xunit;
+    using Xunit;
+
+    public class UseNHibernateSessionTests
+    {
+        [Fact]
+        public async Task Execute_Queryable()
+        {
+            // arrange
+            ;
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/UseNHibernateSessionTests.cs:24]:
```suggestion
```

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/UseNHibernateSessionTests.cs [hunk 27]
```diff
@@ -0,0 +1,462 @@
...
+                    .BuildServiceProvider();
+
+            IRequestExecutor executor =
+                await services.GetRequiredService<IRequestExecutorResolver>()
+                    .GetRequestExecutorAsync();
+
+   
+            using (ISession session = services.GetRequiredService<ISession>())
+            {
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/UseNHibernateSessionTests.cs:43]:
```suggestion
```

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/UseNHibernateSessionTests.cs [hunk 28]
```diff
@@ -0,0 +1,462 @@
...
+                    .GetRequestExecutorAsync();
+
+         
+
+            using (ISession session = services.GetRequiredService<ISession>())
+            {
+                IList<Author> authors = session.Query<Author>().ToList();
+                authors.Add(new Author {Name = "foo"});
+
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/UseNHibernateSessionTests.cs:84]:
```suggestion

```

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/UseNHibernateSessionTests.cs [hunk 29]
```diff
@@ -0,0 +1,462 @@
...
+                foreach (Author author in authors) await session.SaveAsync(author).ConfigureAwait(false);
+            }
+
+
```
Comment (michaelstaib) [inline src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/UseNHibernateSessionTests.cs:88]:
```suggestion

```

File: .build/Build.Parameters.cs [hunk 30]
```diff
@@ -0,0 +1,9 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+partial class Build
```
[the remark this answers is not in the document]
Comment (satishviswanathan) [reply .build/Build.Parameters.cs]: Removed

File: src/HotChocolate/Data/src/NHibernate/Extensions/NHibernateFactory.cs (hunk 2 above)
Comment (satishviswanathan) [reply src/HotChocolate/Data/src/NHibernate/Extensions/NHibernateFactory.cs]: My bad. I've removed those.

File: src/HotChocolate/Data/src/NHibernate/PublicAPI.Shipped.txt [hunk 31]
```diff
@@ -0,0 +1,18 @@
+#nullable enable
+HotChocolate.Data.NHibernateEnumerableExtensions
```
[the remark this answers is not in the document]
Comment (satishviswanathan) [reply src/HotChocolate/Data/src/NHibernate/PublicAPI.Shipped.txt]: Got it. Done

File: src/HotChocolate/Data/src/NHibernate/publish.cmd (hunk 7 above)
Comment (satishviswanathan) [reply src/HotChocolate/Data/src/NHibernate/publish.cmd]: Removed this filed

File: src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/AuthorMap.cs (hunk 11 above)
Comment (satishviswanathan) [reply src/HotChocolate/Data/test/HotChoclate.Data.NHibernate.Tests/AuthorMap.cs]: Done

**Comment by satishviswanathan** [comment]
Thank you all for the help, with the resolver compiler I was able to integrate NHibernate

**Pull request closed without merging**